app/src/main/java/org/mozilla/fenix/tabstray/browser/TabsTouchHelper.kt

```diff
@@ -0,0 +1,132 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.tabstray.browser
+
+import android.graphics.Canvas
+import android.graphics.drawable.Drawable
+import androidx.appcompat.content.res.AppCompatResources
+import androidx.recyclerview.widget.ItemTouchHelper
+import androidx.recyclerview.widget.ItemTouchHelper.ACTION_STATE_IDLE
+import androidx.recyclerview.widget.RecyclerView
+import mozilla.components.browser.tabstray.TabTouchCallback
+import mozilla.components.concept.tabstray.TabsTray
+import mozilla.components.support.base.observer.Observable
+import mozilla.components.support.ktx.android.content.getColorFromAttr
+import mozilla.components.support.ktx.android.content.getDrawableWithTint
+import mozilla.components.support.ktx.android.util.dpToPx
+import org.mozilla.fenix.R
+import org.mozilla.fenix.ext.settings
+import org.mozilla.fenix.home.sessioncontrol.SwipeToDeleteCallback
+
+/**
+ * A callback for consumers to know when a [RecyclerView.ViewHolder] is about to be touched.
+ * Return false if the default behaviour should be ignored.
+ */
+typealias OnViewHolderTouched = (RecyclerView.ViewHolder) -> Boolean
+
+/**
+ * An [ItemTouchHelper] for handling tab swiping to delete.
+ *
+ * @param onViewHolderTouched See [OnViewHolderTouched].
+ */
+class TabsTouchHelper(
+    observable: Observable<TabsTray.Observer>,
+    onViewHolderTouched: OnViewHolderTouched = { true },
+    delegate: Callback = TouchCallback(observable, onViewHolderTouched)
+) : ItemTouchHelper(delegate)
+
+/**
+ * An [ItemTouchHelper.Callback] for drawing custom layouts on [RecyclerView.ViewHolder] interactions.
+ *
+ * @param onViewHolderTouched invoked when a tab is about to be swiped. See [OnViewHolderTouched].
+ */
+class TouchCallback(
+    observable: Observable<TabsTray.Observer>,
+    private val onViewHolderTouched: OnViewHolderTouched
+) : TabTouchCallback(observable) {
+
+    override fun getMovementFlags(
+        recyclerView: RecyclerView,
+        viewHolder: RecyclerView.ViewHolder
+    ): Int {
+        if (!onViewHolderTouched.invoke(viewHolder)) {
+            return ItemTouchHelper.Callback.makeFlag(ACTION_STATE_IDLE, 0)
+        }
+
+        return super.getMovementFlags(recyclerView, viewHolder)
+    }
+
+    override fun onChildDraw(
+        c: Canvas,
+        recyclerView: RecyclerView,
+        viewHolder: RecyclerView.ViewHolder,
+        dX: Float,
+        dY: Float,
+        actionState: Int,
+        isCurrentlyActive: Boolean
+    ) {
+        super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
+
+        if (recyclerView.context.settings().gridTabView) {
```

Comment:
I took this code from the old package, but I think we can refactor this bit out.

Reply:
This is done. We now have a `onViewHolderDraw` callback that delegates it to the app.